src/main/java/controller/BookController.java

```diff
@@ -0,0 +1,339 @@
+     * @param readLink       String representation of the link where the book can be read.
+     * @param downloadLink   String representing the link where the book can be downloaded from.
+     * @return Message describing whether the operation was successful or not.
+     */
+    public String addDownloadableEBook(String bookISBN, String bookTitle, String summary,
+                                       List<Author> authors, List<BookGenre> bookGenres, List<BookCategory> bookCategories,
+                                       String readLink, String downloadLink) {
+        if (isBookValid(bookISBN, bookTitle, summary, authors, bookGenres, bookCategories) && isLinkValid(readLink) && isLinkValid(downloadLink)) {
+            DownloadableEBook newDownloadableEBook = new DownloadableEBook(bookISBN, bookTitle, summary, authors, bookGenres, bookCategories,
+                    readLink, downloadLink);
+
+            bookRepository.addBookToLibrary(newDownloadableEBook);
+
+            return "Book: " + bookTitle + " added successfully to library.";
+        }
+
+        return "Adding book failed.";
+    }
+
+    /**
+     * Validates the provided input by calling methods to validate each parameter separately.
+     * If all are valid creates an EBook and adds it to the repository.
+     *
+     * @param bookISBN       String representation of unique book identifier.
+     * @param bookTitle      Name of the book.
+     * @param summary        Short info about what the book is about.
+     * @param authors        A list of one or more authors.
+     * @param bookGenres     A list of one or more book genres.
+     * @param bookCategories A list of one or more book categories.
+     * @param readLink       String representation of the link where the book can be read.
+     * @return Message describing whether the operation was successful or not.
+     */
+    public String addEBook(String bookISBN, String bookTitle, String summary,
+                           List<Author> authors, List<BookGenre> bookGenres, List<BookCategory> bookCategories,
+                           String readLink) {
+        if (isBookValid(bookISBN, bookTitle, summary, authors, bookGenres, bookCategories) && isLinkValid(readLink)) {
+            EBook newEBook = new EBook(bookISBN, bookTitle, summary, authors, bookGenres, bookCategories,
+                    readLink);
+
+            bookRepository.addBookToLibrary(newEBook);
+
+            return "Book: " + bookTitle + " added successfully to library.";
+        }
+
+        return "Adding book failed.";
+    }
+
+    /**
+     * Validates the link with a regular expression for links.
+     *
+     * @param bookLink String representation of a link.
+     * @return true if the link matches the regexp or false if doesn't.
+     */
+    private boolean isLinkValid(String bookLink) {
+        return Pattern.matches("^(http)(s)*://(www.)*([a-z0-9]+.)+[a-z]+(:[0-9]{1,4})*", bookLink);
```

Comment:
http[s]?://(?:www)?[a-zA-Z0-9]+\\.[a-z]\\/?[a-zA-Z0-9\\/]*

Reply:
In this regex only one .(dot) is allowed.